Sky light implementation

robske110 opened the pull request:
#### If you want to discuss this on a bigger scale please refer to this [thread](https://forums.pmmp.io/threads/implementing-skylight-into-pmmp.19/).
Hello there! This is my first real contribution to pmmp.
# Unfinished?
This is currently awaiting reviews&testing. **If you want to help out** please **test this** using the instructions in this post, or do a **code review**.
Also chunk generation time population is still missing.

Currently this also has massive debug in it, that will be removed if this is ready to merge.

### Known bugs
- Currently all bugs 🐞 🐜  are fixed.

### Missing logic&features
- [x] ~directSkyLight is missing Light Diffusion support~ DONE in https://github.com/pmmp/PocketMine-MP/pull/160/commits/175a5462d1c49acb4bca67b1d7afe01cd90fd4af
- [x] ~skyLightResistance is probably wrong for some blocks [*HELP* would be appreciated here]~ DONE in https://github.com/pmmp/PocketMine-MP/pull/160/commits/175a5462d1c49acb4bca67b1d7afe01cd90fd4af
- [x] ~Support for light ways that have a step in them~ DONE in https://github.com/pmmp/PocketMine-MP/pull/160/commits/11568389135404d0d2aed2cacbd48d9d88a30158
```
      _______
_____|   ____
________|
```
- [x] ~General abort logic in getSkyLightViaWays (details in TODO list below)~

### Todo
- [ ] fix any bugs if found
- [ ] remove debug messages / ~change some to logger->debug~
- [ ] Make performance better
- [x] implement missing features
- [x] remove some unneded variables in getSkyLightViaWays
- [x] Better abort logic
- [x] direct: getHeightMap -> can help to determine if skylight can reach calcpos at all [Problem with half transparent blocks!] DONE in https://github.com/pmmp/PocketMine-MP/pull/160/commits/175a5462d1c49acb4bca67b1d7afe01cd90fd4af
- [x] clean up code
- ~ways: if cannot reach origin ->more research&thinking needed~

### Problems:
- Performance Currently takes 0.001 to 0.05s. Currently due to missing skylight info on chunk population and massive debug messages.
- ~Abort techniques: When will they decrease the performance too much, when will they increase performance~
- Needing to update a 15 block radius if 1 block changed?

# How can I test this?
You can test this using [this](https://github.com/pmmp/PocketMine-MP/files/858131/SkylightTester.php.zip) plugin. Install it, and run the server using https://github.com/robske110/pmmp-PocketMine-MP/tree/skylight_curr. Then just touch a block and you will get the skylight value of it.
If you found a bug, please enable debug and post the console output with a screenshot of the block you calculated for here.

# Wtf is skylight?
A form of light that comes from the sky and can travel horizontally on the ground. It is used for many gameplay mechanics like grass spreading,  spawning. It also affects the brightness of blocks in a certain way.
it is not useable for things like lightdetectors, skylight (at least in MCPC) does not decrease at night. This is done by the client through Brightness, pretty much at the Render level.

# Technical details
This code checks at first for the skylight from above. This is done by going from the origin block to the build limit/the highest block in the chunk and subtracting the resistance of each block in the air from 15. We immediately abort calculating if the block is above the highest block in the chunk. Air has no (vertical) resistance. If the direct skylight is 14 or 15, we save that value and are done here. (We can't get higher SkyLight horizontal) If it's below that we check if we have horizontal ways for the light to get in. We start into a big recursive function that basically looks at each side of the block, and from each of these blocks it checks if it receives direct-skylight, if it is too low/it doesn't recieve skylight it looks if it has ways on the 4 sides or above, and checks these blocks.  Waylight is the light which comes from the four sides for a block.
<!--- https://github.com/pmmp/PocketMine-MP/files/776963/SkylightTester.php.zip --->

robske110 committed:
first work here

robske110 committed:
progress?

robske110 committed:
stuff

robske110 committed:
cleanup && theoretically almost done

robske110 committed:
throw everything around

robske110 committed:
Make it run...

robske110 committed:
i need sleep

robske110 committed:
this 'works' now

and no more freezes

robske110 committed:
fix 2 freezes, fix some bugs, add even more debug

robske110 committed:
fix bugs introduced in last commit *facepalm*

robske110 committed:
polish, add typehints, add phpdoc

dktapps commented on src/pocketmine/level/Level.php:
```diff
@@ -1230,12 +1230,9 @@ public function rayTraceBlocks(Vector3 $pos1, Vector3 $pos2, $flag = false, $fla
-				$level = max($chunk->getBlockLight($pos->x & 0x0f, $pos->y & 0x7f, $pos->z & 0x0f));
-			}
+		if($chunk instanceof FullChunk){
+			//Do not decrease skylightlvl over day. MCPC and MCPE don't do this either.
```
I don't understand the blah about lighting curves because my brain isn't properly awake, but http://minecraft.gamepedia.com/Light#Brightness seems like a must-read.

dktapps commented on src/pocketmine/block/Block.php:
```diff
@@ -279,6 +279,37 @@ public static function init(){
+	 *
+	 * @return int 0-15
+	*/
+	public static function getSkyLightResistance($blockID){
```
... why didn't you use `Block::$lightFilter`? It has correct values already.

Also bear in mind that blocks don't specifically have a _sky_ light filter, just a light filter overall. Read http://minecraft.gamepedia.com/Light#Light-filtering_blocks

robske110 replied:
gonna double check that

robske110 replied:
fixed

dktapps commented on src/pocketmine/block/Block.php:
```diff
@@ -279,6 +279,37 @@ public static function init(){
+	 *
+	 * @return int 1-15
+	*/
+	public static function getHorizontalSkyLightResistance($blockID){
```
Blocks do not have different horizontal/vertical light filters. This looks like a hack.

robske110 replied:
horizontal spread skylight light is always decreasing by one. This is the fastest and cleanest way to implement this probably...

robske110 committed:
Implement Block diffusion & improve performance

Improved performance for blocks that are above the highest block in a
chunk

dktapps commented on src/pocketmine/block/Block.php:
```diff
@@ -279,6 +279,31 @@ public static function init(){
+	 *
+	 * @return int 0-15
+	*/
+	public static function getSkyLightResistance($blockID): int{
```
It's not skylight resistance, it's general light filtering.

robske110 replied:
yea, the function name is probably confusing

robske110 replied:
fixed

robske110 committed:
Fix wrong diffusion

robske110 committed:
OOP for Diffusion, Fix diffusion, refactor resistance=>light filter

robske110 commented:
rebased onto mcpe-1.0

robske110 commented:
Travis failed due to Jenkins 502 error

robske110 commented:
rebased. Going to see if I finally get time to finish this.

robske110 committed:
Skylight: Fixes&Cleanup; Update Heightmap on setBlock

-Rename getSkyLightVerticalFilter to getVerticalSkyLightFilter
-Fix $lightFilter having transparent blocks listed as filtering 1
-getHorizontalLightFilter Fix function call on static member (facepalm)
-getDirectSkylight php’s dumb operator precedence
-Fix max(): empty array
-Fix findlightways
-Fix blocklight freeze (because of the corrected $lightFilter)
-Fix Heightmap not updating onblockset

robske110 commented:
Hello there!
I'd really like someone to help me testing this. Just read under "How can I test this?" in the inital post. There is only one missing feature now, which is also described above.

robske110 committed:
Cleanup; Implement more efficient abort technique

robske110 committed:
Lightways with light from above.

This is it? I can’t find any difference to PC now.

robske110 commented:
Again, testing would be great. This should be in line with MCPC behaviour now, and hopefully also mcpe. Code reviews are also welcome.

dktapps commented on src/pocketmine/level/Level.php:
```diff
@@ -1222,12 +1222,9 @@ public function rayTraceBlocks(Vector3 $pos1, Vector3 $pos2, $flag = false, $fla
-			}
+		if($chunk instanceof Chunk){
+			//Do not decrease skylightlvl over day. MCPC and MCPE don't do this either.
+			$level = max($chunk->getBlockLight($pos->x & 0x0f, $pos->y & Level::Y_MASK, $pos->z & 0x0f), $chunk->getBlockSkyLight($pos->x & 0x0f, $pos->y & Level::Y_MASK, $pos->z & 0x0f));
```
The Y mask is not needed, and I don't recommend using it, as it will produce unexpected wraparound values.

jasonwynn10 commented:
![img_0711](https://cloud.githubusercontent.com/assets/16521025/24086700/e1300fe0-0cea-11e7-9473-e1a7709749c0.PNG)
It appears that your tester plugin breaks the client...

robske110 commented:
I quote
>Please note that due to the high spam of debug messages clients may get unresponsive due to the sheer amount of chat messages.
But i never saw that 😮 (What OS are you using?) 
I always tested with iOS
Well, thanks for trying to test anyways! Really awesome! You could try to ignore it and simply read the values from the console, and I'll change the broadcastMessage to Logger->debug later today.
If you find any bugs apart from Mojang's chat bugs in my skylight implementation, feel free to report.

jasonwynn10 commented:
There was an issue that kept appearing in Level. Block::getVerticalLightFilter needs to be changed to Block::getVerticalSkyLightFilter

robske110 commented:
Oh, looks like i didn't commit that fix correctly. Thanks for the heads-up. Did you experience any other issues (especially with the calculated lightlevels?)

jasonwynn10 commented:
![img_0715](https://cloud.githubusercontent.com/assets/16521025/24119139/9400c79a-0d86-11e7-80dc-3921be26fea5.PNG)
This block is receiving direct sky light, but it still says it's at 0

jasonwynn10 commented:
![screenshot 28](https://cloud.githubusercontent.com/assets/16521025/24124677/4bbf59a0-0d9b-11e7-974a-505cad5a5921.png)
Another error given is a missing element in a max() function in Level

robske110 commented:
I thought i fixed that :/
To your earlier post: going to look into it
Again, thanks for testing, you might want to pause for now until I've fixed those bugs :)

robske110 committed:
Fix small bugs; make debug less annoying

robske110 commented:
@jasonwynn10 I (hopefully) fixed your bugs 🙂 It would be nice if you can test it again. I also updated the SkyLightTester plugin, so you might want to get the new version, i updated the link in the initial post. I also now provide https://github.com/robske110/pmmp-PocketMine-MP/tree/skylight_curr for easier testing. (Simply this merged onto current master)

dktapps commented on src/pocketmine/level/Level.php:
```diff
@@ -1223,8 +1223,8 @@ public function getFullLight(Vector3 $pos) : int{
 		if($chunk instanceof Chunk){
-			//Do not decrease skylightlvl over day. MCPC and MCPE don't do this either.
-			$level = max($chunk->getBlockLight($pos->x & 0x0f, $pos->y & Level::Y_MASK, $pos->z & 0x0f), $chunk->getBlockSkyLight($pos->x & 0x0f, $pos->y & Level::Y_MASK, $pos->z & 0x0f));
+			//Do not decrease skylightlvl over time. MCPC and MCPE don't do this either.
```
Why is this comment necessary at all? It just looks messy.

robske110 replied:
It exists because i simply modified https://github.com/pmmp/PocketMine-MP/blob/master/src/pocketmine/level/Level.php#L1234. But now that you say it, I probably should just get rid of it.

jasonwynn10 commented:
![img_0718](https://cloud.githubusercontent.com/assets/16521025/24150057/091a56b0-0e1b-11e7-825b-2df29acbfbd6.PNG)
That light level should be 13, but after placing the block, it says it's 0. It only gave me the correct level after tapping on it.
Also, when I place or break a block, I have to wait 2 seconds for the server to catch up.

jasonwynn10 commented on src/pocketmine/level/Level.php:
```diff
@@ -1341,23 +1341,23 @@ private function getDirectSkyLight($x, $y, $z){
 	 * @param $currWayResistance   The horizontal resistance of the current root way
```
Can you also correct this PHPdoc comment? It's going to bug me 😅

robske110 commented:
I can't quite understand what's happening there. Could you describe how to place blocks to reproduce?

robske110 commented:
Oh, don't be confused by directSkyLightLevels. Only look at FinalSkyLightLevels.

dktapps commented on src/pocketmine/level/Level.php:
```diff
@@ -1278,8 +1275,148 @@ public function updateAllLight(Vector3 $pos){
+		
+		$directSkyLight = 15;
+		for($i = $y; $i <= $highestBlock; $i++){
+			$lightResistance = Block::getVerticalSkyLightFilter($chunk->getBlockId($x & 0x0f, $i & 0x7f, $z & 0x0f));
```
Better remove the 0x7f here.

dktapps commented on src/pocketmine/level/Level.php:
```diff
@@ -1394,8 +1531,9 @@ public function setBlock(Vector3 $pos, Block $block, bool $direct = false, bool
-		if($this->getChunk($pos->x >> 4, $pos->z >> 4, true)->setBlock($pos->x & 0x0f, $pos->y & Level::Y_MASK, $pos->z & 0x0f, $block->getId(), $block->getDamage())){
+		
+		$chunk = $this->getChunk($pos->x >> 4, $pos->z >> 4, true);
+		if($chunk->setBlock($pos->x & 0x0f, $pos->y & Level::Y_MASK, $pos->z & 0x0f, $block->getId(), $block->getDamage())){
```
and the Y mask here too.

dktapps commented:
This pull request has been obsoleted by the referenced commit above (dab73d8).

Since MCPE 1.1 now actually renders shadows produced by sky light in multiplayer, this was far easier to test with 1.1. Sky light population became critical as of 1.1 due to horrible lighting issues with previously-generated worlds.

My main complaint about this pull request is that it spends a significant amount of CPU time to update the sky-light value of a _single block_. There are potentially a lot more blocks than just the target that need their sky-light levels updating if a block changed. This pull request does not cater for that.

I'm sorry to have to reject this, but since sky light updating is critical to both performance and gameplay as of 1.1, this is not fit for purpose.
